engine/src/main/java/io/zeebe/engine/processing/common/EventHandle.java

```diff
@@ -120,6 +124,35 @@ public void activateElement(
     }
   }
 
+  private boolean isElementActivated(final ExecutableFlowElement catchEvent) {
+    return catchEvent.getElementType() == BpmnElementType.INTERMEDIATE_CATCH_EVENT
+        || catchEvent.getElementType() == BpmnElementType.RECEIVE_TASK
+        || catchEvent.getElementType() == BpmnElementType.EVENT_BASED_GATEWAY;
+  }
+
+  private boolean isInterrupting(final ExecutableFlowElement catchEvent) {
+    if (catchEvent instanceof ExecutableCatchEvent) {
+      final var event = (ExecutableCatchEvent) catchEvent;
+      return event.isInterrupting();
+    } else {
+      return false;
+    }
+  }
+
+  private void activateCatchEvent(
+      final ExecutableFlowElement catchEvent,
+      final long eventScopeKey,
+      final ProcessInstanceRecord elementRecord) {
+
+    elementRecord
+        .setBpmnElementType(catchEvent.getElementType())
+        .setElementId(catchEvent.getId())
+        .setFlowScopeKey(eventScopeKey);
+
+    // TODO (saig0): pass message variables for output mappings (temp/as local vars) (#6187)
+    commandWriter.appendNewCommand(ProcessInstanceIntent.ACTIVATE_ELEMENT, elementRecord);
```

Comment:
So, not sure if this is relevant or not. For ServiceTasks I ran into problems here.

This could happen for a non-interrupting timer boundary event. 

* `CatchEventBehavior.subscribeToTimeEvent()` will write the element instance key of the service task into the record
* Consequently, this line would try to activate the service task element, which is wrong

Maybe there is a bug in the implementation of Service Task. And maybe I got confused along the way. Currently, I don't see why this wouldn't happen for `ReceiveTasks` as well given that they use the same mechanism to subscribe to events.

Reply:
Yes, the timer record references the element instance key of the service task but also the element id of the boundary event (i.e. the handle node id). 

The activate command is written for the element that is referenced by the timer (i.e. the handle node id). In the lines above (147-150), the command is filled with the data from the referenced event (i.e. the boundary event).